src/rgw/driver/rados/rgw_sal_rados.cc

```diff
@@ -2628,6 +2734,112 @@ int RadosObject::transition_to_cloud(Bucket* bucket,
+  read_op->params.lastmod = &read_mtime;
+
+  ret = read_op->prepare(y, dpp);
+  if (ret < 0) {
+    ldpp_dout(dpp, 0) << "XXXXX: read_op failed ret=" << ret << dendl;
+    return ret;
+  }
+
+  map<string, bufferlist> attrs = get_attrs();
+  RGWRados::Object op_target(store->getRados(), bucket->get_info(), *rados_ctx, get_obj());
+  RGWRados::Object::Write obj_op(&op_target);
+	Object* obj = (Object*)this;
+
+  bufferlist bl;
+  auto attr_iter = attrs.find(RGW_ATTR_RESTORE_TYPE);
+  if (attr_iter != attrs.end()) {
+    using ceph::decode;
+    rgw::sal::RGWRestoreType restore_type;
+    decode(restore_type, attr_iter->second);
+    if (restore_type == rgw::sal::RGWRestoreType::Temporary) {
+      attr_iter = attrs.find(RGW_ATTR_MANIFEST);
+      if (attr_iter != attrs.end()) {
+        RGWObjManifest m;
+        try {
+          using ceph::decode;
+          decode(m, attr_iter->second);
+          set_atomic();
+          obj_op.meta.modify_tail = true;
+          obj_op.meta.flags = PUT_OBJ_CREATE;
+          obj_op.meta.category = RGWObjCategory::CloudTiered;
+          obj_op.meta.delete_at = real_time();
+          bufferlist blo;
+          obj_op.meta.data = &blo;
+          obj_op.meta.if_match = NULL;
+          obj_op.meta.if_nomatch = NULL;
+          obj_op.meta.user_data = NULL;
+          obj_op.meta.zones_trace = NULL;
+          obj_op.meta.delete_at = real_time();
+          obj_op.meta.set_mtime = read_mtime;
+
+          RGWObjManifest *pmanifest;
+          RGWObjManifest manifest;
+          pmanifest = &manifest;
+
+	        Object* head_obj = (Object*)this;
+          RGWObjTier tier_config;
+          m.get_tier_config(&tier_config);
+	
+          rgw_placement_rule target_placement(pmanifest->get_head_placement_rule(), tier_config.name);
+
+          pmanifest->set_head(target_placement, head_obj->get_obj(), 0);
+          pmanifest->set_tail_placement(target_placement, head_obj->get_obj().bucket);
+          pmanifest->set_obj_size(0);
+          obj_op.meta.manifest = pmanifest;
+
+          // erase restore attrs
+          attrs.erase(RGW_ATTR_RESTORE_STATUS);
+          attrs.erase(RGW_ATTR_RESTORE_TYPE);
+          attrs.erase(RGW_ATTR_RESTORE_TIME);
+          attrs.erase(RGW_ATTR_RESTORE_EXPIRY_DATE);
+          attrs.erase(RGW_ATTR_CLOUDTIER_STORAGE_CLASS);
+
+          bufferlist bl;
+          bl.append(tier_config.name);
+          attrs[RGW_ATTR_STORAGE_CLASS] = bl;
+
+          const req_context rctx{dpp, y, nullptr};
+          return obj_op.write_meta(0, 0, attrs, rctx, head_obj->get_trace());
+        } catch (const buffer::end_of_buffer&) {
+          // ignore empty manifest; it's not cloud-tiered
+        } catch (const std::exception& e) {
+        }
+      }
+      return 0;
+    }
+  }
+  // object is not restored/temporary; go for regular deletion
+  obj->set_atomic();
+  ret = obj->delete_object(dpp, null_yield, rgw::sal::FLAG_LOG_OP);
```

Comment:
Sorry I don't understand why do we need to delete the object?

Reply:
The original object_expirer code is to delete the object. I added this new routine to handle temp. restored objects as well and it default to delete_object for all other cases.